### PR TITLE
Fix for loop side effects

### LINKED ISSUE #8032: [yul] optimizer's load resolver does not take side effects of user-defined function containing for loop(s) into account
## Description

```
{
        function foo(x) {
                // codecopy alters contents of mem[0]
                for {} x { x := mload(0) codecopy(0, 77, 32)} {}
        }
        // mem[0] = 1337
        mstore(0, 1337)
        // Function with side-effects that alters memory
        foo(42)
        // mem[0] = 1337 may no longer be valid here
        sstore(0, mload(0))
}
```

is optimized to

```
            let _1 := 1337      
            mstore(0, _1)       
            let x := 42                                                
            for { }             
            x                   
            {       
                x := mload(0)
                codecopy(0, 77, 32)
            }
            { }
            // mload(0) is incorrectly resolved to initial assignment
            sstore(0, _1)
        }
```

## Environment

- Compiler version: latest develop

## Steps to Reproduce

```
$ solc --strict-assembly --optimize <test.yul>
```

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/8032

### CODE REVIEW

**Comment by bshastry** [comment]
Abi full suite test is failing. Perhaps, expectation needs to be updated.

**Comment by bshastry** [comment]
Missing bug list entry for `0.5.14`. This may be fixed if this PR is rebased on top of #8034

**Comment by chriseth** [comment]
Bug list entry will be part of the 0.5.15 release (the other PR).

File: test/libyul/yulOptimizerTests/fullSuite/abi_example1.yul [hunk 1]
```diff
@@ -477,15 +477,16 @@
...
-//         if slt(sub(_3, length), 128) { revert(_1, _1) }
-//         let offset := calldataload(add(length, 64))
-//         let _4 := 0xffffffffffffffff
-//         if gt(offset, _4) { revert(_1, _1) }
-//         let value2 := abi_decode_t_array$_t_uint256_$dyn_memory_ptr(add(length, offset), _3)
-//         let offset_1 := calldataload(add(length, 0x60))
-//         if gt(offset_1, _4) { revert(_1, _1) }
-//         let value3 := abi_decode_t_array$_t_array$_t_uint256_$2_memory_$dyn_memory_ptr(add(length, offset_1), _3)
-//         sstore(calldataload(length), calldataload(add(length, 0x20)))
+//         let _4 := mload(0x20)
```
Comment (chriseth) [inline test/libyul/yulOptimizerTests/fullSuite/abi_example1.yul:480]: The main difference is that it re-loads `length` here and does not reuse it like above.

**Review by bshastry — APPROVED** [review]
LGTM. The additional test cases will get merged in the backport PR in any case.

**Comment by chriseth** [comment]
Yes, closing this in favour of #8039

**Pull request closed without merging**